### PR TITLE
feat: support unixfs metadata and formatting it

### PR DESCRIPTION
**Pull request by achingbrain**
Adds `mode` and `mtime` properties to normalised `.add` inputs, also adds functions for turning metadata into strings for CLI use.

BREAKING CHANGE:

In order to support metadata on intermediate directories, `globSource` in this module will now emit directories and files where previously it only emitted files.

### CODE REVIEW

File: src/files/glob-source.js [hunk 1]
```diff
@@ -77,15 +107,29 @@ async function * toGlobSource ({ path, type, prefix }, options) {
...
     cwd: path,
-    nodir: true,
+    nodir: false,
     realpath: false,
     absolute: true
   })
 
   for await (const p of glob(path, '**/*', globOptions)) {
+    if (preserveMode || preserveMtime) {
+      const stat = await fs.stat(p)
```
Comment (alanshaw) [inline src/files/glob-source.js]: In the future we might get some perf wins by [parallel mapping](https://www.npmjs.com/package/streaming-iterables#parallelmap) these.

File: src/files/glob-source.js [hunk 2]
```diff
@@ -77,15 +107,29 @@ async function * toGlobSource ({ path, type, prefix }, options) {
...
+        mtime = parseInt(stat.mtimeMs / 1000)
+      }
+    }
+
     yield {
       path: toPosix(p.replace(prefix, '')),
-      content: fs.createReadStream(p)
+      content: fs.createReadStream(p),
+      mode,
+      mtime
```
Comment (alanshaw) [inline src/files/glob-source.js]: Directories get the same `mode` and `mtime` as files?

File: src/files/glob-source.js [hunk 3]
```diff
@@ -77,15 +107,29 @@ async function * toGlobSource ({ path, type, prefix }, options) {
...
+
+      if (options.preserveMtime) {
+        mtime = parseInt(stat.mtimeMs / 1000)
+      }
+    }
+
     yield {
       path: toPosix(p.replace(prefix, '')),
-      content: fs.createReadStream(p)
+      content: fs.createReadStream(p),
```
Comment (alanshaw) [inline src/files/glob-source.js]: Now that `nodir: false`, will this try to read a directory?

File: src/files/glob-source.js (hunk 2 above)
Comment (achingbrain) [reply src/files/glob-source.js]: They get nothing, unless `mode` or `mtime` are specified as options, in which case they that, unless `preserveMode` or `preserveMtime` are specified, in which case they get whatever they had on disk.

There was a bug here where files only got the `mode` or `mtime` of their containing folder but it is now squashed.

File: src/files/glob-source.js (hunk 3 above)
Comment (achingbrain) [reply src/files/glob-source.js]: Yes, although not now cos it was an bug and has been squashed with tests added.

File: src/files/glob-source.js (hunk 1 above)
Comment (achingbrain) [reply src/files/glob-source.js]: Possibly - our bottlenecks typically aren't from reading paths from the filesystem but it's worth measuring before jumping to conclusions.  Like I just did.  :trollface:

**Comment by achingbrain** [comment]
@hugomrdias if you're happy with this PR could you merge & release it please?

**Comment by hugomrdias** [comment]
can you fix the conflicts 🙏

**Comment by achingbrain** [comment]
@hugomrdias done.

**Comment by achingbrain** [comment]
@hugomrdias this went out in the same release as #15 so now UnixFSv1.5 metadata is dependent on the removal of node streams & pull-streams from IPFS, which it shouldn't be.

Can this PR either be backported to `0.4.x` or can you do a `0.6.0` release with just this PR and `0.7.0` with this PR and #15 please?

**Pull request merged**